### ISSUE TITLE
to/fromNMS(Class<? extends Entity>) incompatible; custom pathfinder wrapper not decodable

### ISSUE THREAD

**Issue by datatags** [GamerCoder215/MobChip#14]
### Bug Type

Compatibility Issues, Console Errors

### Description

I've found two issues here that I'd like to discuss before attempting a fix for them.

The first issue concerns a pair of functions ([fromNMS](https://github.com/GamerCoder215/MobChip/blob/b412f91e05ab06b6f20caa358155a319eba9cbb2/1_19_R1/src/main/java/me/gamercoder215/mobchip/abstraction/ChipUtil1_19_R1.java#L1589-L1648) and [toNMS](https://github.com/GamerCoder215/MobChip/blob/b412f91e05ab06b6f20caa358155a319eba9cbb2/1_19_R1/src/main/java/me/gamercoder215/mobchip/abstraction/ChipUtil1_19_R1.java#L159-L169)) to convert between Bukkit entity classes and NMS entity classes. The issue I'm experiencing is that `fromNMS` converts from NMS classes to Bukkit classes, while `toNMS` actually converts from CraftBukkit classes to NMS classes (only CraftBukkit classes have `getHandle()`, the Bukkit interfaces do not.) This means that MobChip is unable to write pathfinders that use entity parameters (because `toNMS` ends up getting called with the same classes `fromNMS` returns.)

Looking at potential solutions for converting Bukkit entity classes to NMS entity classes, I'm wondering about using a BiMap for entity classes whose names are different in Bukkit and NMS, and then using something like `getOrDefault` to grab the ones that are the same using reflection. (That's sort of like how `toNMS` handles it now, except it uses a big switch statement with the mappings and a default case for the ones that are the same.) This way, we could use the same BiMap for `toNMS` and `fromNMS`.

---

The second issue is about how custom pathfinders are written back to the entity's list of pathfinders. When reading or writing pathfinders to NMS, if the pathfinder isn't a known type, it's wrapped in an anonymous class. (This seems like a potential issue since it seems reading and writing the pathfinders repeatedly would result in a chain of wrapper classes [memory leak], but that's not the main concern here.)

The main reason this is an issue is that when building a `CustomPathfinder` from a `Goal`, it uses reflection to find the target entity of the pathfinder, but the wrapper for building a `Goal` from a `CustomPathfinder` does not contain an entity field because it just passes all function calls directly to the original `CustomPathfinder`.

I was considering a solution like making a real, non-anonymous class extending Goal like `CustomGoal` that provides access to the original `CustomPathfinder`. Then, before we wrap a custom `Goal`, we can check if it's a `CustomGoal`, and if it is, just grab the original `CustomPathfinder` from it and use that instead of a wrapper. (Sorry if this is confusing 😕)

Even if we do this, it's probably a good idea to keep the ability to wrap `Goal`s into `CustomPathfinder`s in case another plugin adds custom pathfinders to an entity using just NMS.

### Expected Behavior

(I'm sure you can guess the answers to this but it's in the issue template 🙂 )

I'd expect MobChip to be able to properly read and re-write pathfinders that concern entities.

---

I'd expect MobChip to be able to read its own custom pathfinders after writing them.

### Actual Behavior

```
[18:05:46 ERROR]: org.bukkit.entity.HumanEntity.getHandle()
[18:05:46 ERROR]: java.base/java.lang.Class.getMethod(Class.java:2259)
[18:05:46 ERROR]: MobChipTest.jar//me.gamercoder215.mobchip.abstraction.ChipUtil1_19_R1.toNMS(ChipUtil1_19_R1.java:162)
[18:05:46 ERROR]: MobChipTest.jar//me.gamercoder215.mobchip.abstraction.ChipUtil1_19_R1.toNMS(ChipUtil1_19_R1.java:265)
[18:05:46 ERROR]: MobChipTest.jar//me.gamercoder215.mobchip.abstraction.ChipUtil1_19_R1.addPathfinder(ChipUtil1_19_R1.java:400)
[18:05:46 ERROR]: MobChipTest.jar//me.gamercoder215.mobchip.abstraction.ChipUtil.addPathfinders(ChipUtil.java:53)
[18:05:46 ERROR]: MobChipTest.jar//me.gamercoder215.mobchip.bukkit.BukkitAI.updateAI(BukkitAI.java:43)
[18:05:46 ERROR]: MobChipTest.jar//me.gamercoder215.mobchip.bukkit.BukkitAI.add(BukkitAI.java:113)
[18:05:46 ERROR]: MobChipTest.jar//me.gamercoder215.mobchip.bukkit.BukkitAI.put(BukkitAI.java:70)
[18:05:46 ERROR]: MobChipTest.jar//me.datatags.mobchiptest.MobChipTest.onCommand(MobChipTest.java:29)
```

---

```
java.lang.IllegalArgumentException: Entity cannot be null
        at be.isach.ultracosmetics.shaded.mobchip.ai.goal.Pathfinder.<init>(Pathfinder.java:24) ~[UltraCosmetics-2.9-RELEASE.jar:?]
        at be.isach.ultracosmetics.shaded.mobchip.ai.goal.CustomPathfinder.<init>(CustomPathfinder.java:16) ~[UltraCosmetics-2.9-RELEASE.jar:?]
        at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1$4.<init>(ChipUtil1_19_R1.java:1813) ~[UltraCosmetics-2.9-RELEASE.jar:?]
        at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.custom(ChipUtil1_19_R1.java:1813) ~[UltraCosmetics-2.9-RELEASE.jar:?]
        at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.fromNMS(ChipUtil1_19_R1.java:1942) ~[UltraCosmetics-2.9-RELEASE.jar:?]
        at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.lambda$getGoals$0(ChipUtil1_19_R1.java:136) ~[UltraCosmetics-2.9-RELEASE.jar:?]
        at java.lang.Iterable.forEach(Iterable.java:75) ~[?:?]
        at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.getGoals(ChipUtil1_19_R1.java:136) ~[UltraCosmetics-2.9-RELEASE.jar:?]
        at be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitAI.updateMap(BukkitAI.java:36) ~[UltraCosmetics-2.9-RELEASE.jar:?]
        at be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitAI.<init>(BukkitAI.java:31) ~[UltraCosmetics-2.9-RELEASE.jar:?]
        at be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitBrain.getGoalAI(BukkitBrain.java:57) ~[UltraCosmetics-2.9-RELEASE.jar:?]
        at be.isach.ultracosmetics.cosmetics.pets.Pet.onEquip(Pet.java:93) ~[UltraCosmetics-2.9-RELEASE.jar:?]
```

**Comment by GamerCoder215** [comment]
Thanks for the (detailed) explanation!

I've left (what I believe to be) some reasoning in #13, if you could take a look at that.
After I'm done reviewing your PR, I'll merge it when I get the chance. Thanks again for your help!

**Comment by GamerCoder215** [comment]
Patched in [`1129017`](https://github.com/GamerCoder215/MobChip/commit/1129017).

Left some comments to explain a couple of things- such as the fact that (a lot) of classes are exactly the same name as bukkit ones, or just need to add an `Entity` prefix to the Class Name, which can save a lot of lines.

Marking as fixed.
